### ISSUE TITLE
RocksDB : hash index slows down COLLECT query

### ISSUE THREAD

**Issue by matcho** [arangodb/arangodb#3552]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.2

Mode:
- [x] Single-Server

Storage-Engine:
- [x] rocksdb

On this operating system:
  - [x] Ubuntu .deb

### this is an AQL-related issue:
I'm issuing AQL via:
- [x] arangosh

We have a collection with ~500 000 documents.
```
127.0.0.1:8529@mydb> db.users.count();
485391
```

Issuing a COLLECT query on `username` field takes ~1.3s
```
127.0.0.1:8529@mydb> db._query('FOR u IN users COLLECT un = u.username WITH COUNT INTO nb FILTER nb > 1 SORT nb DESC RETURN { un, nb }').getExtra();
{ 
  "stats" : { 
    "writesExecuted" : 0, 
    "writesIgnored" : 0, 
    "scannedFull" : 485391, 
    "scannedIndex" : 0, 
    "filtered" : 351751, 
    "httpRequests" : 0, 
    "executionTime" : 1.2599263191223145 
  }, 
  "warnings" : [ ] 
}
```

Adding a "hash" index (but if I understand correctly, indexes are all the same type in RocksDB) on `username` slows down the same query to ~2.8s
```
127.0.0.1:8529@mydb> db.users.ensureIndex({ type: "hash", fields: [ "username" ] });
{ 
  "deduplicate" : true, 
  "fields" : [ 
    "username" 
  ], 
  "id" : "users/11662800", 
  "isNewlyCreated" : true, 
  "selectivityEstimate" : 0.9250225835591689, 
  "sparse" : false, 
  "type" : "hash", 
  "unique" : false, 
  "code" : 201 
}

127.0.0.1:8529@mydb> db._query('FOR u IN users COLLECT un = u.username WITH COUNT INTO nb FILTER nb > 1 SORT nb DESC RETURN { un, nb }').getExtra();
{ 
  "stats" : { 
    "writesExecuted" : 0, 
    "writesIgnored" : 0, 
    "scannedFull" : 0, 
    "scannedIndex" : 485391, 
    "filtered" : 351751, 
    "httpRequests" : 0, 
    "executionTime" : 2.8389158248901367 
  }, 
  "warnings" : [ ] 
}
```

Explain suggests that the index is used for sorting only
```
127.0.0.1:8529@mydb> db._explain('FOR u IN users COLLECT un = u.username WITH COUNT INTO nb FILTER nb > 1 SORT nb DESC RETURN { un, nb }').getExtra();
Query string:
 FOR u IN users COLLECT un = u.username WITH COUNT INTO nb FILTER nb > 1 SORT nb DESC RETURN { un, nb 
 }

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
 11   IndexNode         485391     - FOR u IN users   /* hash index scan */
  3   CalculationNode   485391       - LET #3 = u.`username`   /* attribute expression */   /* collections used: u : users */
  4   CollectNode       388312       - COLLECT un = #3 WITH COUNT INTO nb   /* sorted */
  5   CalculationNode   388312       - LET #5 = (nb > 1)   /* simple expression */
  6   FilterNode        388312       - FILTER #5
  7   SortNode          388312       - SORT nb DESC
  8   CalculationNode   388312       - LET #7 = { "un" : un, "nb" : nb }   /* simple expression */
  9   ReturnNode        388312       - RETURN #7

Indexes used:
 By   Type   Collection   Unique   Sparse   Selectivity   Fields           Ranges
 11   hash   users        false    false        92.50 %   [ `username` ]   *

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-index-for-sort
  6   move-calculations-down
```

But still, it seems strange to me.
Is this an expected behaviour ?

Thanks,
Mathias

**Comment by jsteemann** [comment]
Reproduce test case:
```js
(function() {
  db._drop("users"); 
  db._create("users"); 
  var n = 500 * 1000; 
  for (var i = 0; i < n; ++i) { 
    db.users.insert({ username: "user" + Math.floor(Math.random() * n * 20) });
  }
  var q = `FOR u IN users COLLECT un = u.username WITH COUNT INTO nb FILTER nb > 1 SORT nb DESC RETURN { un, nb }`; 
  var time = require("internal").time;

  console.log("no index");
  db._explain(q);  
  var s = time(); 
  db._query(q);
  console.log("time:", time() - s);

  db.users.ensureIndex({ type: "hash", fields: ["username"]});

  console.log("with index");
  db._explain(q);  
  s = time(); 
  db._query(q);
  console.log("time:", time() - s);
})();
```

**Comment by jsteemann** [comment]
The query with the hash index takes about twice as long with the above test data.
The reason is that the COLLECT method is "sorted" with the index, and not "hash". The "sorted" COLLECT algorithm is obviously slower than the hash-based COLLECT.

**Comment by matcho** [comment]
Thanks for your answer.

Is it right to say that, in terms of results, the "sorted" COLLECT has no advantage over the "regular" COLLECT ? If this is true, could the query optimizer decide not to use the index in such a case ?

By the way `statement.explain()` accepts optimizer rules to modify the plan, for example `stmt.explain({ optimizer: { rules: ["-all"] } })`.
But how can I pass optimizer rules to query execution, so that the query is executed with this modified plan ? This could be a solution in the case above.

**Comment by jsteemann** [comment]
The optimizer rules can be passed to query explanation and execution in a similar fashion, by passing the `options` attribute into `db._createStatement()`:

**explain**
```js
stmt = db._createStatement({ 
  query: queryString, 
  options: { optimizer: { rules: ["-all"] } } 
}); 
plan = stmt.explain();
```
**execute**
```js
stmt = db._createStatement({ 
  query: queryString, 
  options: { optimizer: { rules: ["-all"] } } 
}); 
result = stmt.execute();
```
Another alternative is to use the even more simplified versions 
* `db._explain(queryString, bindVars, options);` or 
* `db._query(queryString, bindVars, options);`

**Comment by jsteemann** [comment]
In general it makes sense that the query optimizer has the choice between the index-based COLLECT and the hash table-based COLLECT. Both versions have different trade-offs: the former only will look at one item/group at a time while the later needs to keep the entire result in memory.
However, in this case it looks like the sorted COLLECT variant is a lot slower, and this needs investigation.

**Comment by jsteemann** [comment]
With MMFiles, the sorted variant of COLLECT is only slightly slower than the hash table-based variant.

**Comment by dothebart** [comment]
Confirming that the freshly released [ArangoDB 3.7](https://www.arangodb.com/2020/08/arangodb-3-7-a-big-step-forward-for-multi-model/) still reproduces this behavior.

**Comment by matcho** [comment]
Late update: now that collection has 6M documents, using the index is much faster. Guess this problem only occurs for small number of documents.
